library/src/cache/patch_manager.rs

```diff
@@ -0,0 +1,883 @@
+                self.patches_state.next_boot_patch = Some(last_boot_patch);
+            } else {
+                self.patches_state.last_booted_patch = None;
+
+                if let Err(e) = self.delete_patch_artifacts(last_boot_patch.number) {
+                    // If the last booted patch is no longer bootable, delete its artifacts.
+                    error!(
+                        "Failed to delete patch artifacts for patch {}. Error: {}",
+                        last_boot_patch.number, e
+                    );
+                }
+            }
+        }
+    }
+}
+
+impl ManagePatches for PatchManager {
+    fn add_patch(&mut self, patch_number: usize, file_path: &Path) -> Result<()> {
+        if !file_path.exists() {
+            bail!("Patch file {} does not exist", file_path.display());
+        }
+
+        let patch_path = self.patch_artifact_path(patch_number);
+
+        std::fs::create_dir_all(self.patch_dir(patch_number))
+            .with_context(|| format!("create_dir_all failed for {}", patch_path.display()))?;
+
+        std::fs::rename(file_path, &patch_path)?;
+
+        let new_patch = PatchMetadata {
+            number: patch_number,
+            size: std::fs::metadata(&patch_path)?.len(),
+        };
+
+        // If a patch was never booted (next_boot_patch != last_booted_patch), we should delete
+        // it here before setting next_boot_patch to the new patch.
+        if let (Some(last_boot_patch), Some(next_boot_patch)) = (
+            self.patches_state.next_boot_patch,
+            self.patches_state.last_booted_patch,
+        ) {
+            if last_boot_patch.number != next_boot_patch.number {
+                if let Err(e) = self.delete_patch_artifacts(next_boot_patch.number) {
+                    error!(
+                        "Failed to delete patch artifacts for patch {}. Error: {}",
+                        patch_number, e
+                    );
+                }
+            }
+        }
+
+        self.patches_state.next_boot_patch = Some(new_patch);
+        self.patches_state.highest_seen_patch_number = self
+            .patches_state
+            .highest_seen_patch_number
+            .map(|highest_patch_number: usize| highest_patch_number.max(patch_number))
+            .or(Some(patch_number));
+        self.save_patches_state()
+    }
+
+    fn last_successfully_booted_patch(&self) -> Option<PatchInfo> {
+        self.patches_state
+            .last_booted_patch
+            .map(|patch| self.patch_info_for_number(patch.number))
+    }
+
+    fn next_boot_patch(&mut self) -> Option<PatchInfo> {
+        let next_boot_patch = match self.patches_state.next_boot_patch {
+            Some(patch) => patch,
+            None => return None,
+        };
+
+        if let Err(e) = self.validate_patch_is_bootable(&next_boot_patch) {
+            error!("Patch {} is not bootable: {}", next_boot_patch.number, e);
+
+            self.try_fall_back_to_last_booted_patch();
```

Comment:
Does this need to report back that it was successful/not in falling back? I guess it's job is to leave us in a consistent state of things bootable?

Reply:
I think we generally need some mechanism of reporting that we decided a patch wasn't bootable. Noodling on that now